File: packages/java/src/main/java/dev/kreuzberg/config/LayoutDetectionConfig.java
package dev.kreuzberg.config;

import java.util.HashMap;
import java.util.Map;

/**
 * Layout detection configuration for document layout analysis.
 *
 * @since 4.4.0
 */
public final class LayoutDetectionConfig {
	private final String preset;
	private final Double confidenceThreshold;
	private final Boolean applyHeuristics;

	private LayoutDetectionConfig(Builder builder) {
		this.preset = builder.preset;
		this.confidenceThreshold = builder.confidenceThreshold;
		this.applyHeuristics = builder.applyHeuristics;
	}

	public static Builder builder() {
		return new Builder();
	}

	public String getPreset() {
		return preset;
	}

	public Double getConfidenceThreshold() {
		return confidenceThreshold;
	}

	public Boolean getApplyHeuristics() {
		return applyHeuristics;
	}

	public Map<String, Object> toMap() {
		Map<String, Object> map = new HashMap<>();
		if (preset != null) {
			map.put("preset", preset);
		}
		if (confidenceThreshold != null) {
			map.put("confidence_threshold", confidenceThreshold);
		}
		if (applyHeuristics != null) {
			map.put("apply_heuristics", applyHeuristics);
		}
		return map;
	}

	public static final class Builder {
		private String preset = "fast";
		private Double confidenceThreshold;
		private Boolean applyHeuristics = true;

		private Builder() {
		}

		public Builder preset(String preset) {
			this.preset = preset;
			return this;
		}

		public Builder confidenceThreshold(Double confidenceThreshold) {
			this.confidenceThreshold = confidenceThreshold;
			return this;
		}

		public Builder applyHeuristics(Boolean applyHeuristics) {
			this.applyHeuristics = applyHeuristics;
			return this;
		}

		public LayoutDetectionConfig build() {
			return new LayoutDetectionConfig(this);
		}
	}

	static LayoutDetectionConfig fromMap(Map<String, Object> map) {
		if (map == null) {
			return null;
		}
		Builder builder = builder();
		if (map.get("preset") instanceof String) {
			builder.preset((String) map.get("preset"));
		}
		Object confidenceThresholdValue = map.get("confidence_threshold");
		if (confidenceThresholdValue instanceof Number) {
			builder.confidenceThreshold(((Number) confidenceThresholdValue).doubleValue());
		}
		Object applyHeuristicsValue = map.get("apply_heuristics");
		if (applyHeuristicsValue instanceof Boolean) {
			builder.applyHeuristics((Boolean) applyHeuristicsValue);
		}
		return builder.build();
	}
}

File: packages/php/src/Config/LayoutDetectionConfig.php
<?php

declare(strict_types=1);

namespace Kreuzberg\Config;

/**
 * Layout detection configuration.
 *
 * Controls document layout analysis including region detection presets,
 * confidence filtering, heuristic post-processing, and table structure
 * recognition model selection.
 *
 * @example
 * ```php
 * use Kreuzberg\Config\LayoutDetectionConfig;
 *
 * $layout = new LayoutDetectionConfig(
 *     preset: 'accurate',
 *     confidenceThreshold: 0.75,
 *     applyHeuristics: true,
 *     tableModel: 'tatr',
 * );
 * ```
 */
readonly class LayoutDetectionConfig
{
    /**
     * @param string $preset Model preset controlling accuracy vs speed trade-off.
     *                       Supported values: "fast", "accurate". Default "accurate".
     * @param float|null $confidenceThreshold Minimum confidence threshold for detected layout
     *                                        regions (0.0-1.0). Regions below this threshold are discarded.
     *                                        Default null (use engine default).
     * @param bool $applyHeuristics Whether to apply heuristic post-processing to refine
     *                              layout regions. Default true.
     * @param string|null $tableModel Table structure recognition model to use.
     *                                Supported values: "tatr", "slanet_wired", "slanet_wireless", "slanet_plus", "slanet_auto".
     *                                Default null (use engine default).
     */
    public function __construct(
        public string $preset = 'accurate',
        public ?float $confidenceThreshold = null,
        public bool $applyHeuristics = true,
        public ?string $tableModel = null,
    ) {
    }

    /**
     * Create configuration from array data.
     *
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        $preset = $data['preset'] ?? 'accurate';
        $confidenceThreshold = isset($data['confidence_threshold']) ? (float) $data['confidence_threshold'] : null;
        $applyHeuristics = isset($data['apply_heuristics']) ? (bool) $data['apply_heuristics'] : true;
        $tableModel = isset($data['table_model']) ? (string) $data['table_model'] : null;

        return new self(
            preset: is_string($preset) ? $preset : 'accurate',
            confidenceThreshold: $confidenceThreshold,
            applyHeuristics: $applyHeuristics,
            tableModel: $tableModel,
        );
    }

    /**
     * Convert configuration to array for FFI.
     *
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        $result = [
            'preset' => $this->preset,
            'apply_heuristics' => $this->applyHeuristics,
        ];

        if ($this->confidenceThreshold !== null) {
            $result['confidence_threshold'] = $this->confidenceThreshold;
        }

        if ($this->tableModel !== null) {
            $result['table_model'] = $this->tableModel;
        }

        return $result;
    }
}
